### PR TITLE
Feature/custom emoji and tagline views

### PR DESCRIPTION
**Pull request by Freakazoid182**
Add paged endpoints for `custom_emoji` and `tagline`, as per #4577

Does not remove the resources from `GetSiteResponse`.

@dessalines:
* How to handle breaking changes to `GetSiteResponse`? Can `custom_emojis` and `taglines` only be removed on `api/v4`?
* Taglines are now created, updated and deleted through the `site` update endpoint. Do you want complete separation of this resource, like with `custom_emoji`?

### CODE REVIEW

File: crates/db_views/src/tagline_view.rs
Comment (Nutomic) [inline crates/db_views/src/tagline_view.rs:34]: Use iterator to avoid clone, something like `taglines.into_iter().map(|tagline| TaglineView { tagline }).collect()`. 

Honestly Im not sure a view makes any sense here, as it doesnt contain any extra data so we might as well use `Tagline` directly.

File: crates/api_crud/src/custom_emoji/list.rs
Comment (Nutomic) [inline crates/api_crud/src/custom_emoji/list.rs:23]: I think `.context()` makes more sense here, [example](https://github.com/LemmyNet/lemmy/blob/main/crates/federate/src/worker.rs#L133).

**Comment by Nutomic** [comment]
> How to handle breaking changes to GetSiteResponse? Can custom_emojis and taglines only be removed on api/v4?

We can publish this in 0.19.x, and then remove the `GetSiteResponse` fields in 0.20.0. That way api clients have some time to do the migration.

**Comment by dessalines** [comment]
> How to handle breaking changes to GetSiteResponse? Can custom_emojis and taglines only be removed on api/v4?

I've added the breaking change label, so we'll wait till after this upcoming release to merge this. Its unavoidably a breaking change tho, because taglines and emojis are going to come back with separate endpoints.

@Nutomic I'd rather just push this to 0.20

File: crates/api_common/src/custom_emoji.rs
Comment (dessalines) [inline crates/api_common/src/custom_emoji.rs:65]: @makotech222 look into your emoji library, because these may need a search term, optional category, etc.

File: crates/api_crud/src/tagline/list.rs
Comment (dessalines) [inline crates/api_crud/src/tagline/list.rs:23]: Look at the other endpoints for how these errors are handled.

File: crates/db_views/src/structs.rs
Comment (dessalines) [inline crates/db_views/src/structs.rs:202]: This is a pointless wrapper.

File: crates/db_views/src/tagline_view.rs
Comment (dessalines) [inline crates/db_views/src/tagline_view.rs:22]: Order by the published date descending, so it shows newest first.

**Comment by Freakazoid182** [comment]
> > How to handle breaking changes to GetSiteResponse? Can custom_emojis and taglines only be removed on api/v4?
> 
> I've added the breaking change label, so we'll wait till after this upcoming release to merge this. Its unavoidably a breaking change tho, because taglines and emojis are going to come back with separate endpoints.
> 
> @Nutomic I'd rather just push this to 0.20

Confirming: This means we'll remove the `custom_emojis` and `taglines` from GetSiteResponse in this PR right?

**Comment by dessalines** [comment]
yes. IMO a single random tagline can could back with getsiteresponse also, so that front ends can use a random tagline. But proper editing of taglines would be the new endpoints.

File: crates/api_common/src/custom_emoji.rs
Comment (Freakazoid182) [reply crates/api_common/src/custom_emoji.rs:65]: I've added a category search parameter. This column is also already indexed

**Comment by Freakazoid182** [comment]
> yes. IMO a single random single tagline can could back with getsiteresponse also, so that front ends can use a random tagline. But proper editing of taglines would be the new endpoints.

Understood. I'll also need to add the create, update and delete endpoints for taglines then indeed.

File: crates/db_views/src/tagline_view.rs
Comment (Freakazoid182) [reply crates/db_views/src/tagline_view.rs]: Removed `TaglineView`, it was unnecessary indeed.

File: crates/api_crud/src/custom_emoji/list.rs
Comment (Freakazoid182) [reply crates/api_crud/src/custom_emoji/list.rs]: Removed the message

File: crates/api_crud/src/tagline/list.rs
Comment (Freakazoid182) [reply crates/api_crud/src/tagline/list.rs]: Removed the message

File: crates/db_views/src/structs.rs
Comment (Freakazoid182) [reply crates/db_views/src/structs.rs]: Removed `TaglineView`

File: crates/db_views/src/tagline_view.rs
Comment (Freakazoid182) [reply crates/db_views/src/tagline_view.rs]: `TaglineView` was removed. New `list` function can be found [here](https://github.com/LemmyNet/lemmy/pull/4580/files#diff-4118e2f1f35bfd0d48a41dabd4366577863e150820ea6ab5372e73a004271e95R68)

**Comment by Freakazoid182** [comment]
> yes. IMO a single random single tagline can could back with getsiteresponse also, so that front ends can use a random tagline. But proper editing of taglines would be the new endpoints.

I've added new endpoints to edit taglines, and `GetSiteResponse` returns a random tagline. Please check if you're happy with the solution of getting a single random tagline from the database. Testing it with a handful of taglines seems to work fine.

**Review by dessalines — COMMENTED** [review]
Looking good so far! 

Also can you mark this as a draft PR? I'm mainly afraid of one of us prematurely merging it, since it has some breaking changes. After this upcoming release, we'll be doing breaking changes again, and one of us can mark it as ready for review.

File: crates/db_schema/src/impls/tagline.rs
Comment (Freakazoid182) [reply crates/db_schema/src/impls/tagline.rs]: Good points, it looks like this should also be done for `custom_emoji` then

File: src/api_routes_http.rs
Comment (Freakazoid182) [reply src/api_routes_http.rs]: They've been moved under `/admin` :+1:

File: crates/db_schema/src/impls/tagline.rs
Comment (Freakazoid182) [reply crates/db_schema/src/impls/tagline.rs]: Also implemented `Crud` on `CustomEmoji` now

File: crates/api_crud/src/tagline/update.rs
Comment (dessalines) [inline crates/api_crud/src/tagline/update.rs:26]: One other thing I realized I missed, both for tagline update and create. `content` is a markdown field, so it needs to run some extra functions. Check out `crates/api_crud/src/comment/update.rs` , the `content` field, but you need:

`process_markdown_opt(&data.body, &slur_regex, &url_blocklist, &context).await?;`

and 

`is_valid_body_field`

Comment (Freakazoid182) [reply crates/api_crud/src/tagline/update.rs]: Thanks, would've totally missed that. Looks like there wasn't any validation in the previous solution, so good that it's being added now.

File: crates/db_schema/src/source/tagline.rs
Comment (dessalines) [inline crates/db_schema/src/source/tagline.rs:37]: Should be required.

File: crates/api_crud/src/tagline/create.rs
Comment (Freakazoid182) [reply crates/api_crud/src/tagline/create.rs]: I created `is_valid_tagline_content`, which is possibly a bit overkill, but I thought it couldn't hurt to have specific validation rules around taglines.

File: crates/utils/src/utils/validation.rs
Comment (Nutomic) [inline crates/utils/src/utils/validation.rs:26]: This seems extremely long, same as max post body. I think 200 or so would be enough.

Comment (Nutomic) [reply crates/utils/src/utils/validation.rs:26]: Btw only `is_valid_post_title()` calls `trim()` before length check. Would probably make sense to `trim` inside of `min_length_check` and `min_length_check` so that it applies to all strings.

cc @dessalines

File: crates/utils/src/error.rs
Comment (Nutomic) [inline crates/utils/src/error.rs:169]: Just use a single error `TaglineInvalid`. Not worth making separate errors which have to be translated into dozens of languages.

File: crates/db_views/src/custom_emoji_view.rs
Comment (Nutomic) [inline crates/db_views/src/custom_emoji_view.rs:75]: This is unnecessary because all emojis have the same local site id. Honestly you can remove the `local_site_id` column, it doesnt seem to have any use.

File: crates/utils/src/utils/validation.rs
Comment (SleeplessOne1917) [reply crates/utils/src/utils/validation.rs:26]: Hexbear uses taglines far longer than 200 characters. Lemmygrad as well.

File: crates/db_schema/src/impls/tagline.rs
Comment (SleeplessOne1917) [inline crates/db_schema/src/impls/tagline.rs:65]: Is there a potential edge case bug if filtering is done after the limit? Also, why match on a local site ID? Taglines should only be relevant in the context of the local site, so it's not like you have to worry about filtering out taglines from federated instances.

File: crates/db_views/src/custom_emoji_view.rs
Comment (SleeplessOne1917) [reply crates/db_views/src/custom_emoji_view.rs:75]: I made a similar point regarding taglines.

**Review by SleeplessOne1917 — CHANGES_REQUESTED** [review]

File: crates/utils/src/utils/validation.rs
Comment (Freakazoid182) [reply crates/utils/src/utils/validation.rs:26]: I took 50000 which is the default max body length. Seems like too much indeed. Maybe 500 chars?

File: crates/db_schema/src/source/tagline.rs
Comment (Freakazoid182) [reply crates/db_schema/src/source/tagline.rs:45]: This is because of a [comment](https://github.com/LemmyNet/lemmy/pull/4580/files/f1b993f90ed262b20e03bde4ed4233afd11695d0#r1553942482) by @dessalines that the column is a nullable.

File: crates/db_schema/src/impls/tagline.rs
Comment (Freakazoid182) [reply crates/db_schema/src/impls/tagline.rs]: `local_site_id` has been removed from the `Tagline` resource

File: crates/db_schema/src/source/tagline.rs
Comment (SleeplessOne1917) [reply crates/db_schema/src/source/tagline.rs]: Why would it need two options to make it nullable? Shouldn't one suffice?

Comment (Freakazoid182) [reply crates/db_schema/src/source/tagline.rs]: Is it possibly to be able to set it to `Some(None)`, effectively setting the `updated` column to `NULL`. Where as if you would have a `Option<DateTime<Utc>>` and set it to `None` it would be ignored (no updated) rather than set to `NULL`?

File: crates/utils/src/utils/validation.rs
Comment (Nutomic) [reply crates/utils/src/utils/validation.rs:26]: I checked hexbear and lemmygrad taglines manually, longest one I noticed is 1170 chars. That seems extremely long but if we want to support it, we can set a limit of 1500 or 2000 chars.

File: crates/api_crud/src/tagline/create.rs
Comment (dessalines) [reply crates/api_crud/src/tagline/create.rs]: Nice, thx.

File: crates/db_schema/src/source/tagline.rs
Comment (dessalines) [reply crates/db_schema/src/source/tagline.rs:35]: Correct, but also it should be `content: Option<String>`

To clarify:


_ | _
--- | ---
Inserts to a NOT NULL column | `String`
Inserts to a Nullable column | `Option<String>`
Updates to a NOT NULL column | `Option<String>`
Updates to a Nullable column | `Option<Option<String>`

**Review by dessalines — COMMENTED** [review]
We also need feedback from @makotech222 , because we need to tie in how async loading of emojis, and searching, will work in the front ends.

File: crates/utils/src/utils/validation.rs
Comment (Freakazoid182) [reply crates/utils/src/utils/validation.rs:26]: Agreed @dessalines. Shall we keep it as is, or remove the max limit completely?

**Comment by Freakazoid182** [comment]
> > We also need feedback from @makotech222 , because we need to tie in how async loading of emojis, and searching, will work in the front ends.
> 
> Like i said, I don't think separating out the emoji list into pages is a worthwhile endeavor. Taglines is fine, because we technically only need a random one on page load and then you can paginate for admin page.
> 
> Its up to you guys how you want to do the frontend side if you paginate emojis. You'd have to make sure Tribute and EmojiMart can both support it. Just remember its going to require a network call + a db call to handle each and every search anytime emojis are used.

I only gave the Tribute and EmojiMart code a quick look. It does seem that for EmojiMart all emojis needs to be loaded when constructing the `Picker`. In any case, we can put back the emojis on the `GetSiteResponse`, or all Emojis need to be retrieved on site render through an separate API call. Or we keep it like now, and another solution will have to be found.

I think having the paged API for custom emojis is fine to have for the admin functionality.

**Comment by dessalines** [comment]
> It does seem that for EmojiMart all emojis needs to be loaded when constructing the Picker. In any case, we can put back the emojis on the GetSiteResponse, or all Emojis need to be retrieved on site render through an separate API call. Or we keep it like now, and another solution will have to be found.

Dang. Well let's allow an override on the paging limits on the new `ListCustomEmojis` endpoint you added, so that other requests to `GetSite` aren't hit with such a massive emoji list. There are a few examples of this override in the code, lmk if you need help finding them.

In the future tho we should look to either add async to emojimart, or find a new library, because fetching thousands of emojis all at once isn't going to scale up well.

File: crates/utils/src/utils/validation.rs
Comment (dessalines) [reply crates/utils/src/utils/validation.rs:26]: As is seems fine to me. You've externalized it so we can always tweak it later easily.

**Review by SleeplessOne1917 — APPROVED** [review]

**Comment by makotech222** [comment]
Just want to note, hexbear probably has the most expansive emoji and tagline list in lemmyverse. Our getsiteresponse in its entirety is about 1.7mb. Not that bad

**Comment by Freakazoid182** [comment]
> There are a few examples of this override in the code, lmk if you need help finding them.

The only thing I could find was on the `LocalImage` db_schema (`get_all_helper`). There's a `ignore_page_limits` bool there. I've added this to the `ListCustomEmojis` query.

**Comment by Freakazoid182** [comment]
I noticed that on the `GetSiteResponse` API, there's a short lived cache to limit calls to the database. It it worth it to add this to `ListCustomEmojis` for requests that use `ignore_page_limits`?

**Comment by Nutomic** [comment]
Yes it would be much simpler with caching. Caching different pages would be possible with cache key `(category, page, limit)`, but an attacker could bypass it by calling with various different parameters. So I think its better to get rid of pagination entirely for taglines and emojis, so that there is only a single possible response which can be cached very well. This also makes it much easier for clients to implement.

**Comment by dessalines** [comment]
Its not just about the fetch size, its also about limiting the amount of info a front end page has to render and display. 

Let's say you're editing emojis and taglines. You don't want to have to render, load images for, and scroll through 800 emojis all at once.

As for caching, I think that's on us to add it if we think its necessary, not you.

File: crates/db_views/src/custom_emoji_view.rs
Comment (dessalines) [inline crates/db_views/src/custom_emoji_view.rs:38]: You can probably get rid of this one now. Or leave it as a helper function, but have it call `self.list(.... true)

File: crates/db_views/src/custom_emoji_view.rs
Comment (dessalines) [inline crates/db_views/src/custom_emoji_view.rs:81]: The category ordering should probably only happen if they supplied a category, just to be faster. So you could add `.order_by(category)` into the `if let Some(category)` above,  and add a `query = query.then_order_by(id)` below that block.

**Comment by Nutomic** [comment]
Keeping pagination is also fine. In that case it makes no sense to do caching in Rust, better leave it to nginx. That is already handled by cache-control headers which Lemmy sends for unauthenticated users.

**Comment by Freakazoid182** [comment]
> As for caching, I think that's on us to add it if we think its necessary, not you.

> In that case it makes no sense to do caching in Rust, better leave it to nginx

Sounds good to me.

> Keeping pagination is also fine. 

I'll keep the the paging as it is then

**Comment by dessalines** [comment]
Its up to us to fix the conflicts, I'll do that soon.

**Comment by dessalines** [comment]
K updated, and this is ready for a review.

**Review by dessalines — APPROVED** [review]
LGTM, will need plenty of front-end work.

**Comment by Nutomic** [comment]
Made a few changes: 
- Removed is_valid_tagline_content() because a limit of 50k is completely meaningless. Anyway taglines can only be set by admins so we dont need much verification
- Made fields TaglineInsertForm and TaglineUpdateForm mandatory (it doesnt make sense to create a tagline without content, or to update a tagline without both content and update time
- move emoji order statement in custom_emoji_view.rs so that it applies to all queries
- add comment for GetSiteResponse.tagline

Good to merge now.

**Review by dessalines — APPROVED** [review]

**Comment by dessalines** [comment]
I'll get conflicts.

**Pull request merged**
